%% Copyright (c) 2011, Pacific Biosciences of California, Inc.

%% All rights reserved.

%% Redistribution and use in source and binary forms, with or without
%% modification, are permitted provided that the following conditions are
%% met: 

%% * Redistributions of source code must retain the above copyright
%% notice, this list of conditions and the following disclaimer.

%% * Redistributions in binary form must reproduce the above copyright
%% notice, this list of conditions and the following disclaimer in the
%% documentation and/or other materials provided with the distribution.

%% * Neither the name of Pacific Biosciences nor the names of its
%% contributors may be used to endorse or promote products derived from
%% this software without specific prior written permission.

%% THIS SOFTWARE IS PROVIDED BY PACIFIC BIOSCIENCES AND ITS CONTRIBUTORS
%% "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
%% LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
%% A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL PACIFIC
%% BIOSCIENCES OR ITS CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
%% INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
%% BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS
%% OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
%% ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
%% TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
%% USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
%% DAMAGE.
\documentclass{article}

%%%%%%%%%%%%%%%%%%%%%%%% Standard Packages %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\usepackage{epsfig}
%\usepackage{graphicx}
%\usepackage{graphics}
%\usepackage{amssymb}
%\usepackage{amsmath}
%\usepackage{fancyvrb}
%\usepackage{comment}
%\usepackage{fancyhdr}
\usepackage{color}
%\usepackage{amsfonts}
%\usepackage{float}
%\usepackage{subfig}
\usepackage{hyperref}

\usepackage{amsmath}
\usepackage{amscd}
\usepackage[tableposition=top]{caption}
\usepackage{ifthen}
\usepackage[utf8]{inputenc}


%%%%%%%%%%%%%%%%%%%%%%%% Adapted from Sweave %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\DefineVerbatimEnvironment{Rcode}{Verbatim}{fontshape=sl, frame=single, 
%  framesep=2mm, fontsize=\small, baselinestretch=.5}

%%%%%%%%%%%%%%%%%%%%%%%% Page and Document Setup %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
\addtolength{\oddsidemargin}{-0.875in}
\addtolength{\topmargin}{-0.875in}
\addtolength{\textwidth}{1.75in}
\addtolength{\textheight}{1.75in}

%\renewcommand{\topfraction}{0.9}        % max fraction of floats at top
%\renewcommand{\bottomfraction}{0.8}     % max fraction of floats at bottom

% Parameters for TEXT pages (not float pages):
%\setcounter{topnumber}{2}
%\setcounter{bottomnumber}{2}
%\setcounter{totalnumber}{4}             % 2 may work better
%\setcounter{dbltopnumber}{2}            % for 2-column pages
%\renewcommand{\dbltopfraction}{0.9}     % fit big float above 2-col. text
%\renewcommand{\textfraction}{0.07}      % allow minimal text w. figs

% Parameters for FLOAT pages (not text pages):
%\renewcommand{\floatpagefraction}{0.7}          % require fuller float pages

% N.B.: floatpagefraction MUST be less than topfraction !!
%\renewcommand{\dblfloatpagefraction}{0.7}       % require fuller float pages


%\def\argmax{\operatornamewithlimits{arg\,max}}
%\def\argmin{\operatornamewithlimits{arg\,min}}

\definecolor{myblue}{rgb}{0.25, 0, 0.75}
\definecolor{mygold}{rgb}{1,0.8,0.2}


\definecolor{gray}{rgb}{0.5, 0.5, 0.5}

\newcommand{\myurl}[1]{\href{http://#1}{\textcolor{gray}{\texttt{#1}}}}
\newcommand{\myem}[1]{\structure{#1}}
\newcommand{\myurlshort}[2]{\href{http://#1}{\textcolor{gray}{\textsf{#2}}}}

\newcommand{\RPackage}[1]{\textcolor{gray}{\textsf{#1}}}
\newcommand{\PL}[1]{\texttt{#1}}
\newcommand{\RCode}[1]{\texttt{#1}}
\newcommand{\RFunction}[1]{\textsf{#1}}
\newcommand{\RClass}[1]{\textcolor{mygold}{\textsf{#1}}}
\newcommand{\BIOCfunction}[1]{\textcolor{orange}{#1}}

%%%%%%%%%%%%%%%%%%%%%%% options for sweave %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%\SweaveOpts{prefix.string=plots/plots, eps=FALSE, echo=FALSE}

%\documentclass{article}


\begin{document}

\title{A Modification Motif Analysis Demo}
\author{Pat Marks}
\maketitle

\section {Introduction}
This document demonstrates a typical analysis of kinetic modification detection data from bacterial genome.
The raw data used here is generated by sequencing native E.Coli K12 on the PacBio RS, and performing secondary analysis using the RS.Modification.Detection workflow. Some words about RM dogma -- decoding the RM system is the goal of this analysis.

\begin{itemize}
\item Load and explore the modification calls from the RS.Modification.Detection workflow
\item Find the sequence motifs that are common to most of the modified genome hits
\item Find instances of those motifs in the genome sequence
\item Determine the what fraction of the motifs instances were detected as methylated
\end{itemize}

\subsection{R Setup}
The analysis workflow demonstrated here on R 2.14 and R 2.15 on Linux and Windows.
From \PL{CRAN} we require \RPackage{ggplot2} and \RPackage{plyr}. From \PL{Bioconductor} we require
\RPackage{Biostrings} and \RPackage{cosmo}. See the appendix for details of download and installation detatils.

Please prepare your R environment as follows.  \RCode{scripts.R} is included with this tutorial.
<<libs>>=
library(ggplot2)
library(plyr)
library(Biostrings)
library(cosmo)
source('scripts.R')
@

\section{Loading modification data}

We start by loading the raw modifications calls from that are produced by SMRTPortal in \verb@modifications.gff.gz@. This can be downloaded from the job results page in SMRTPortal, or accessed directly from the SMRTportal job folder on the file server.  \verb@modifications.gff.gz@ is compliant with \PL{GFFv3} specification (\href{http://www.sequenceontology.org/gff3.shtml}[www.sequenceontology.org]). See Table~\ref{table:gff} for a description of the columns in the \PL{GFF} file. 

\begin{table}[ht]
\begin{center}
\begin{tabular}{ll}
\hline
Column & Description \\
\hline
seqid  &    Reference tag (e.g. ref00001) \\
source	&	  Name of tool -- 'kinModCall' \\
type    &   Modification type -- currently we use a generic tag "modified\_base" \\
start   &   Location of modification \\
end     &   Location of modification plus one \\
score   &   Phred transformed p-value of detection \\
strand  &   Sample strand containing modification \\
phase   &   Not applicable \\
\hline
attributes &  Fields below are packed in the \PL{GFF} attributes column \\
IPDRatio & Ratio between mean IPD of observed data to IPD of unmodified DNA \\
context &  Reference sequence -20bp to +20bp around start, converted to current strand \\
coverage & Number of valid IPD observations at this site \\
\hline
\end{tabular}
\caption{Contents of modifications.gff.gz file}
\end{center}
\label{table:gff}
\end{table}

The included \PL{R} script contains a \PL{GFF} file reader that extracts some extra atrribute columns used by the modification detection tool. Take a look at the data contained in the \PL{GFF} file. The context field contains a 41 base context centered around  the detected modification -- pull out the center base (position 21). Summarize the \RCode{coverage} column of the \PL{GFF} to see how many reads contribute to each modification call. Confident m6A detection generally requires coverage $>20$ per strand.

<<gff>>=
# Load GFF file
gff <- '/mnt/secondary/Smrtanalysis/userdata/jobs/040/040041/data/modifications.gff.gz'
hits <- readModificationsGff(gff)
hits$CognateBase <- substr(hits$context, 21, 21)
head(hits)
summary(hits$coverage)
@

We now make some plots from the \PL{GFF} data to assess the quality and type of the modification calls. Figure~\ref{fig:pp1} shows a histogram of the scores of the GFF entries, coloured by the cognate base.  This will give you a sense of how strong your signals are and whether the strongest signals are enriched on any base. For our E.Coli test genome the predominant modification is 6-methyl adenosine, so most of the significant modification detections are at A positions.

\begin{figure}
\begin{center}
<<label=pp1,fig=TRUE,echo=TRUE>>=
p <- qplot(score, colour=CognateBase, geom='freqpoly', data=hits, binwidth=5)
show(p)
@
\end{center}
\caption{Modification Scores by cognate base}
\label{fig:pp1}
\end{figure}


\begin{figure}
\begin{center}
<<label=pp2,fig=TRUE,echo=TRUE>>=
p <- qplot(coverage, score, colour=CognateBase, alpha=I(0.3), data=hits[1:10000,])
show(p)
@
\end{center}
\caption{Score vs. Coverage}
\label{fig:pp2}
\end{figure}


The histogram in Figure~\ref{fig:pp1} indicates that the interesting A bases have a score cutoff of roughly 45.  We select these hits, then sort in decreasing order of score, so we consider the strongest signal first.

<<filter>>=
goodHits <- subset(hits, score > 45)
goodHits <- goodHits[order(goodHits$score, decreasing=T),]
workHits <- goodHits
@

\section{Motif Finding}

We use the motif finding package \PL{cosmo} (\myurl{http://www.bioconductor.org/packages/release/bioc/html/cosmo.html}) to search for conserved motifs in the DNA sequences surrounding our modification detections.  First, we convert the context snippets into a cosmo compatible format. Next, we select a small number of hits to pass to \PL{cosmo} (it will not return if you provide too many at once). We feed the contexts to cosmo, then look at the resulting logo plot (Figure~\ref{fig:logo1}. The 'GATC' motif is strongly detected. \RPackage{cosmo} searches for motifs anywhere within the supplied sequences. In our case the motif is always appears at the same position within sequence windows we supplied. The motifs hits detected by cosmo should therefore have the same starting position.  We check for this by examining \RCode{table(cosmoOut@motifs@pos)}. Clearly the mode is at position 15. The A in the GATC motif occurs at position 7 in the cosmo motif, which corresponds to $15 + 7 - 1 = 21$, the center, cognate base of the sequences we passed to \RPackage{cosmo}

<<motif>>=
sequenceList <- llply(workHits$context[1:200], function(x) list(seq=as.character(x), desc=''))
cosmoOut <- cosmo(seqs=sequenceList,minW=4, maxW=10, revComp=F, silent=T)

table(cosmoOut@motifs@pos)
startPos <- 15 + 7 - 1
startPos
@

We view the logo plot of the motif found (Figure~\ref{fig:logo})

\begin{figure}
\begin{center}
<<label=logo1,fig=TRUE,echo=TRUE>>=
plot(cosmoOut)
@
\end{center}
\caption{Logo plot - First \RPackage{cosmo} iteration}
\label{fig:logo1}
\end{figure}

We can check which hits match this context using the \RCode{labelContexts} function and see how many GATC hits we have. The \RCode{labelContexts} takes a charactervector of context strings, and a vector of motif strings and their associated methlyated positions. We make a rough approximation of the number of GATC sites we would expect on both stands of a random genome of length $|G|$ as $2|G|/4^n$ where $n$ is the number of bases in the motif.  We see that most GATC are methylated. We will perform a more accurate analysis in the next section. We can now remove the 'GATC' hits from working list. We still have many of unassigned GFF hits, so we can run \RPackage{cosmo} on the reduced list of hits.

<<motif>>=
motif <- 'GATC'
position <- 2
motifLabels <- labelContexts(workHits$context, motif, position)

table(motifLabels)

genomeSize <- max(hits$end)
expectedHits <- 2 * genomeSize / (4^4) 
expectedHits

workHits <- workHits[motifLabels != motif,]
nrow(workHits)
@

We now run \RPackage{cosmo} on the remaining:
<<cosmo2>>=
sequenceList <- llply(workHits$context[1:200], function(x) list(seq=as.character(x), desc=''))
cosmoOut2 <- cosmo(seqs=sequenceList,minW=4, maxW=10, revComp=F, silent=T)

motif <- 'ACNNNNNNGT'
position <- 1
motifLabels <- labelContexts(workHits$context, motif, position)

table(motifLabels)

genomeSize <- max(hits$end)
expectedHits <- 2 * genomeSize / (4^4) 
expectedHits
@

\begin{figure}
\begin{center}
<<label=logo2,fig=TRUE,echo=TRUE>>=
plot(cosmoOut2)
@
\end{center}
\caption{Logo plot: Second \RPackage{cosmo} iteration}
\label{fig:logo2}
\end{figure}


Figure~\ref{fig:logo2} shows the logo plot of the second \RPackage{cosmo} run. It's unclear what's happening now. We have found a highly confident 4 base motif but we observe far less than the expected number. In fact our E.Coli sample contains a pair of methyltransferases that methylate reverse-complementary motifs that share a common 4 base sub-motif. These two motifs occur in equal numbers.  \RPackage{cosmo} was not designed to handle situation like this, so we must continue on by hand, inspecting the hits manually.  Hand inspection of the hits shows that 'GATC', 'GCACNNNNNNGTT', 'AACNNNNNNGTGC' are the fully methylated motifs in our sample.

\section{Genome Annotation}
We can load the genome sequence and annotate it instances of each motif, to determine the genome-wide methylation fraction of our motifs. The \RFunction{genomeAnnotation} function returns a \PL{data.frame} containing one row for each match in the supplied genome of each motif supplied. Genome positions can match multiple motifs, which gives multiples row at the same genome position

<<annotate>>=
# Load the genome sequence
seq_path <- '/mnt/secondary/Smrtanalysis/userdata/references/ecoli/sequence/ecoli.fasta'
dna_seq <- read.DNAStringSet(seq_path)

motifs = c('GATC', 'GCACNNNNNNGTT', 'AACNNNNNNGTGC')
positions = c(2, 3, 2)

genomeAnnotations <- genomeAnnotation(dna_seq, motifs, positions)
head(genomeAnnotations)

table(genomeAnnotations$motif)
@

We merge the \RFunction{genomeAnnotation} output with our \PL{GFF} data.frame to count the number of motif instances that exist in the genome and the number that were detected as methylated. We merge the genomeAnnotation and goodHits tables by genome position and strand, with \RCode{all=T} to include GFF hits that are not annotated with a motif, and genome motif instances that do not have a GFF hit. We adjust the merged data.frame to indicate these cases.

<<merge>>=
goodHits$seqid <- as.integer(substr(goodHits$seqname, 4,11))
mm <- merge(goodHits, genomeAnnotations, all=T)
mm$motif[is.na(mm$motif)] <- 'NoMotif' 
mm$feature[is.na(mm$feature)] <- 'not_detected' 
table(mm$feature, mm$motif)
@

In our goodHits table we have 3032 'modified\_base' calls that do not occur at an annotated genome position.  For genome positions matching the GATC motif, 37728 were present detected ('modified\_base') and 512 were not ('not\_detected'). We can assess whether we are likely to have missed any methylated motifs by comparing the score distributions for the positions matching a motif to the 'NoMotif' set, as in Figure~\ref{fig:scoreByMotif}.

\begin{figure}
\begin{center}
<<label=scoreByMotif,fig=TRUE,echo=TRUE>>=
p <- qplot(score, ..density.., colour=motif, geom='freqpoly', data=subset(mm, feature=='modified_base'), binwidth=3, xlim=c(0,200))
show(p)
@
\end{center}
\caption{Score distribution by motif annotation}
\label{fig:scoreByMotif}
\end{figure}

\appendix

\section{R Package Installation}

\PL{R} can be downloaded for Linux, Mac and Windows from \myurl{http://r-project.org}
We also recommend the graphical front-end \PL{RStudio}, available from \myurl{rstudio.org}.
The following R packages need to be installed to run through this demo.
\RPackage{ggplot2}, \RPackage{plyr} are available through the built-in \PL{R} pacakge manager. Instructions for installing Bioconductor packages is available here: \myurl{www.bioconductor.org/install}.
\RPackage{cosmo} and \RPackage{Biostrings} are required from Bioconductor

\subsection{Session Info}
Here we give information about the version of R and installed packages
used to generate this document. 
<<echo=TRUE>>=
sessionInfo()
@ 
\bigskip
\footnotesize{
  For Research Use Only. Not for use in diagnostic procedures.
  Copyright 2011, Pacific Biosciences of California, Inc. All rights
  reserved. Information in this document is subject to change without
  notice. Pacific Biosciences assumes no responsibility for any errors
  or omissions in this document. Certain notices, terms, conditions
  and/or use restrictions may pertain to your use of Pacific Biosciences
  products and/or third party products. Please refer to the applicable
  Pacific Biosciences Terms and Conditions of Sale and to the applicable
  license terms at http://www.pacificbiosciences.com/licenses.html.
  
  Pacific Biosciences, the Pacific Biosciences logo, PacBio, SMRT and
  SMRTbell are trademarks of Pacific Biosciences in the United States
  and/or certain other countries. All other trademarks are the sole
  property of their respective owners.
}
\end{document}